File: examples/motion_planning/motion_planning_test/src/problem.rs
use argmin::core::{CostFunction, Error, Gradient, Hessian};
use nalgebra as na;
use ndarray::{Array1, Array2};

use crate::gen;

/// Weights for the cost functions
#[derive(Clone)]
pub struct Weights {
    // Weight for the final state cost
    pub final_state: f64,
    // Weight for the dynamics cost
    pub dynamics: f64,
}

#[derive(Clone)]
pub struct ProblemInfo {
    pub dt: f64,
    pub mass: f64,
}

// Modify to change the planning horizon and time discretization of problem
// A larger N will result in a more accurate trajectory, but will take longer to solve
// Remember to keep consistent with `problem.py`
pub const PROBLEM_INFO: ProblemInfo = ProblemInfo {
    dt: 0.25,
    mass: 1.0,
};
pub const HORIZON: f64 = 4.0;
pub const N: usize = (HORIZON / PROBLEM_INFO.dt) as usize;

pub const STATE_DIM: usize = 12;
pub const POS_VEL_ACC_DIM: usize = 9;
pub const TRAJECTORY_DIM: usize = STATE_DIM * N;
// We don't want to optimize the initial state, so we subtract its size
// from the total trajectory size
pub const TRAJECTORY_TO_OPT_DIM: usize = TRAJECTORY_DIM - POS_VEL_ACC_DIM;

pub const MASS: f64 = 1.0;

/// A discrete point along our trajectory
#[derive(serde::Serialize, serde::Deserialize, Clone, Debug)]
pub struct State {
    pub position: na::Vector3<f64>,
    pub velocity: na::Vector3<f64>,
    pub acceleration: na::Vector3<f64>,
    pub control: na::Vector3<f64>,
}

impl State {
    pub fn new(
        position: na::Vector3<f64>,
        velocity: na::Vector3<f64>,
        acceleration: na::Vector3<f64>,
        control: na::Vector3<f64>,
    ) -> Self {
        Self {
            position,
            velocity,
            acceleration,
            control,
        }
    }

    /// Get a flat svector of the state to pass into the generated residuals
    pub fn to_flat(&self) -> na::SVector<f64, STATE_DIM> {
        let mut flat = na::SVector::<f64, STATE_DIM>::zeros();
        flat.fixed_rows_mut::<3>(0).copy_from(&self.position);
        flat.fixed_rows_mut::<3>(3).copy_from(&self.velocity);
        flat.fixed_rows_mut::<3>(6).copy_from(&self.acceleration);
        flat.fixed_rows_mut::<3>(9).copy_from(&self.control);
        flat
    }

    /// Get a state from a flat svector of the state
    pub fn from_flat(flat: &na::SVector<f64, STATE_DIM>) -> Self {
        Self {
            position: flat.fixed_rows::<3>(0).clone_owned(),
            velocity: flat.fixed_rows::<3>(3).clone_owned(),
            acceleration: flat.fixed_rows::<3>(6).clone_owned(),
            control: flat.fixed_rows::<3>(9).clone_owned(),
        }
    }
}

/// The planning problem.
/// We implement the `CostFunction`, `Gradient`, and `Hessian` argmin traits for this struct,
/// where each of the trait functions use a fast generated function from wrenfold.
/// We optimize the trajectory by minimizing the cost function using a trust region method.
#[derive(Clone)]
pub struct Problem {
    // The desired final state of the trajectory
    pub desired_final_state: State,
    // The weights for the cost function
    pub weights: Weights,
}

impl Problem {
    pub fn new(desired_final_state: State, weights: Weights) -> Self {
        Self {
            desired_final_state,
            weights,
        }
    }

    /// Get a flat state from the optimization problem params Array1 at index i
    fn get_state(&self, p: &Array1<f64>, i: usize) -> na::SVector<f64, STATE_DIM> {
        let p = p.as_slice().unwrap();
        let mut state = na::SVector::<f64, STATE_DIM>::zeros();
        let index_offset = i * STATE_DIM;
        state
            .fixed_rows_mut::<3>(0)
            .copy_from_slice(&p[index_offset..index_offset + 3]);
        state
            .fixed_rows_mut::<3>(3)
            .copy_from_slice(&p[index_offset + 3..index_offset + 6]);
        state
            .fixed_rows_mut::<3>(6)
            .copy_from_slice(&p[index_offset + 6..index_offset + 9]);
        state
            .fixed_rows_mut::<3>(9)
            .copy_from_slice(&p[index_offset + 9..index_offset + 12]);
        state
    }

    /// Get a flat trajectory from the optimization problem params Array1
    fn get_trajectory(&self, p: &Array1<f64>) -> na::SVector<f64, TRAJECTORY_DIM> {
        let mut trajectory = na::SVector::<f64, TRAJECTORY_DIM>::zeros();
        for i in 0..N {
            let state = self.get_state(p, i);
            trajectory
                .fixed_rows_mut::<STATE_DIM>(i * STATE_DIM)
                .copy_from(&state.to_owned());
        }
        trajectory
    }
}

impl CostFunction for Problem {
    type Param = Array1<f64>;
    type Output = f64;

    fn cost(&self, p: &Self::Param) -> Result<Self::Output, Error> {
        let mut cost = 0.0;

        let trajectory = self.get_trajectory(p);
        gen::problem_cost(
            &trajectory,
            &self.desired_final_state.to_flat(),
            &self.weights,
            &PROBLEM_INFO,
            &mut cost,
        );
        Ok(cost)
    }
}

impl Gradient for Problem {
    type Param = Array1<f64>;
    type Gradient = Array1<f64>;

    fn gradient(&self, p: &Self::Param) -> Result<Self::Gradient, Error> {
        let mut gradient = na::SVector::<f64, TRAJECTORY_TO_OPT_DIM>::zeros();
        let trajectory = self.get_trajectory(p);
        gen::problem_jacobian(
            &trajectory,
            &self.desired_final_state.to_flat(),
            &self.weights,
            &PROBLEM_INFO,
            &mut gradient,
        );

        let mut gradient_to_use = na::SVector::<f64, TRAJECTORY_DIM>::zeros();
        gradient_to_use
            .fixed_rows_mut::<TRAJECTORY_TO_OPT_DIM>(POS_VEL_ACC_DIM)
            .copy_from(&gradient);
        Ok(Array1::from_vec(gradient_to_use.as_slice().to_vec()))
    }
}

impl Hessian for Problem {
    type Param = Array1<f64>;
    type Hessian = Array2<f64>;

    fn hessian(&self, p: &Self::Param) -> Result<Self::Hessian, Error> {
        let mut hessian = na::SMatrix::<f64, TRAJECTORY_TO_OPT_DIM, TRAJECTORY_TO_OPT_DIM>::zeros();
        let trajectory = self.get_trajectory(p);
        gen::problem_hessian(
            &trajectory,
            &self.desired_final_state.to_flat(),
            &self.weights,
            &PROBLEM_INFO,
            &mut hessian,
        );

        let mut hessian_to_use = na::SMatrix::<f64, TRAJECTORY_DIM, TRAJECTORY_DIM>::zeros();
        hessian_to_use
            .view_mut(
                (POS_VEL_ACC_DIM, POS_VEL_ACC_DIM),
                (TRAJECTORY_TO_OPT_DIM, TRAJECTORY_TO_OPT_DIM),
            )
            .copy_from(&hessian);

        Ok(
            Array2::from_shape_vec(hessian_to_use.shape(), hessian_to_use.as_slice().to_vec())
                .unwrap(),
        )
    }
}

File: examples/motion_planning/problem.py
import argparse
import typing
from dataclasses import dataclass

import wrenfold as wf
from wrenfold import sym, type_annotations, type_info

# Modify to change the planning horizon and time discretization of problem
# A larger N will result in a more accurate trajectory, but will take longer to solve
# Remember to keep consistent with `problem.rs`
# NOTE: Increasing N will slow down the code generation and rust compilation due to generating
# an output hessian that scales quadratically in size.
HORIZON = 4.0
DT = 0.25
N = int(HORIZON / DT)


# A discrete point along our trajectory
# Corresponds to a flat `State` struct in `problem.rs`
# pub struct State {
#     pub position: na::Vector3<f64>,
#     pub velocity: na::Vector3<f64>,
#     pub acceleration: na::Vector3<f64>,
#     pub control: na::Vector3<f64>,
# }
STATE_DIM = 12
State = typing.Annotated[sym.MatrixExpr, type_annotations.Shape(STATE_DIM, 1)]


# A trajectory is a sequence of states
# TODO: When wrenfold supports sequences of custom types,
# a trajectory can be a sequence of `State`s instead of a matrix
Trajectory = typing.Annotated[sym.MatrixExpr, type_annotations.Shape(STATE_DIM * N, 1)]


@dataclass
class Weights:
    """
    Weights for the cost function.
    Associated with the `Weights` struct in `problem.rs`
    """

    final_state: wf.FloatScalar
    dynamics: wf.FloatScalar


@dataclass
class ProblemInfo:
    """
    Information about the problem.
    Associated with the `ProblemInfo` struct in `problem.rs`
    """

    dt: wf.FloatScalar
    mass: wf.FloatScalar


def get_pos(state: State) -> wf.Vector3:
    return state[0:3, :]


def get_vel(state: State) -> wf.Vector3:
    return state[3:6, :]


def get_acc(state: State) -> wf.Vector3:
    return state[6:9, :]


def get_control(state: State) -> wf.Vector3:
    return state[9:12, :]


def dynamics(state: State, mass: wf.FloatScalar, dt: wf.FloatScalar) -> State:
    """
    Given a state, compute the next state after dt seconds
    """
    pos = get_pos(state)
    vel = get_vel(state)
    acc = get_acc(state)
    control = get_control(state)

    return sym.vstack(
        [
            pos + vel * dt + acc * sym.float_constant(0.5) * dt**2,
            vel + acc * dt,
            control * (1.0 / mass),
            control,
        ]
    )


def cost_jacobian(cost: sym.Expr, state: State):
    return sym.vector(cost).jacobian(state).T


def cost_hessian(cost: sym.Expr, state: State):
    return cost_jacobian(cost, state).jacobian(state)


def desired_state_objective(state: State, desired_state: State, weight: wf.FloatScalar):
    """
    Penalize the distance between the current state and the desired state
    """
    cost = weight * (desired_state - state).squared_norm()
    return cost


def dynamics_objective(
    state: State,
    next_state: State,
    weight: wf.FloatScalar,
    mass: wf.FloatScalar,
    dt: wf.FloatScalar,
):
    """
    Penalize the distance between the next state and the state predicted by the dynamics
    """
    cost = weight * (next_state - dynamics(state, mass, dt)).squared_norm()
    return cost


def problem(
    trajectory: Trajectory,
    desired_final_state: State,
    weights: Weights,
    problem_info: ProblemInfo,
) -> sym.Expr:
    """
    Fixed size optimization problem for trajectory optimization
    """
    cost = sym.float_constant(0.0)
    for i in range(0, N):
        index_offset = i * STATE_DIM
        index_next_offset = index_offset + STATE_DIM
        state = trajectory[index_offset:index_next_offset, :]

        # Final state cost
        cost += desired_state_objective(state, desired_final_state, weights.final_state * i**2)

        # Dynamics cost
        if i < N - 1:
            index_next_next_offset = index_next_offset + STATE_DIM
            next_state = trajectory[index_next_offset:index_next_next_offset, :]
            cost += dynamics_objective(
                state, next_state, weights.dynamics, problem_info.mass, problem_info.dt
            )

    return cost


def problem_cost(
    trajectory: Trajectory,
    desired_final_state: State,
    weights: Weights,
    problem_info: ProblemInfo,
):
    """
    Cost function to generate with wrenfold.
    """
    cost = problem(
        trajectory,
        desired_final_state,
        weights,
        problem_info,
    )
    return [
        wf.OutputArg(
            expression=cost,
            name="cost",
        )
    ]


def problem_jacobian(
    trajectory: Trajectory,
    desired_final_state: State,
    weights: Weights,
    problem_info: ProblemInfo,
):
    """
    Cost function jacobian to generate with wrenfold.
    """
    cost = problem(
        trajectory,
        desired_final_state,
        weights,
        problem_info,
    )
    # We don't want to optimize the initial state's pos/vel/acc
    # so we skip them when taking the jacobian
    trajectory_to_optimize = trajectory[9:]
    jacobian = cost_jacobian(cost, trajectory_to_optimize)
    return [
        wf.OutputArg(
            expression=jacobian,
            name="cost_jacobian",
        )
    ]


def problem_hessian(
    trajectory: Trajectory,
    desired_final_state: State,
    weights: Weights,
    problem_info: ProblemInfo,
):
    """
    Cost function hessian to generate with wrenfold.
    """
    cost = problem(
        trajectory,
        desired_final_state,
        weights,
        problem_info,
    )
    # We don't want to optimize the initial state's pos/vel/acc
    # so we skip them when taking the hessian
    trajectory_to_optimize = trajectory[9:]
    hessian = cost_hessian(cost, trajectory_to_optimize)
    return [
        wf.OutputArg(
            expression=hessian,
            name="cost_hessian",
        )
    ]


class CustomRustGenerator(wf.RustGenerator):
    def format_custom_type(self, element: type_info.CustomType) -> str:
        """Place our custom types into the `crate` namespace."""
        if element.python_type in [Weights, ProblemInfo]:
            return f"crate::{element.name}"
        return self.super_format(element)


def main(args: argparse.Namespace):
    code = ""
    for function in [problem_cost, problem_jacobian, problem_hessian]:
        code += wf.generate_function(func=function, generator=CustomRustGenerator())
        code += "\n\n"
    code = CustomRustGenerator.apply_preamble(code)
    wf.mkdir_and_write_file(code=code, path=args.output)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("output", type=str, help="Output path")
    return parser.parse_args()


if __name__ == "__main__":
    main(parse_args())
